src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs

```diff
@@ -34,29 +34,60 @@ public override BoundNode VisitDeconstructionAssignmentOperator(BoundDeconstruct
         /// </summary>
         private BoundExpression RewriteDeconstruction(BoundTupleExpression left, Conversion conversion, BoundExpression right, bool isUsed)
         {
+            var lhsTemps = ArrayBuilder<LocalSymbol>.GetInstance();
+            var lhsEffects = ArrayBuilder<BoundExpression>.GetInstance();
+            ArrayBuilder<Binder.DeconstructionVariable> lhsTargets = GetAssignmentTargetsAndSideEffects(left, lhsTemps, lhsEffects);
+            BoundExpression result = RewriteDeconstruction(lhsTargets, conversion, left.Type, right, isUsed);
+            Binder.DeconstructionVariable.FreeDeconstructionVariables(lhsTargets);
+            if (result is null)
+            {
+                lhsTemps.Free();
+                lhsEffects.Free();
+                return null;
+            }
+
+            return _factory.Sequence(lhsTemps.ToImmutableAndFree(), lhsEffects.ToImmutableAndFree(), result);
+        }
+
+        private BoundExpression RewriteDeconstruction(
+            ArrayBuilder<Binder.DeconstructionVariable> lhsTargets,
+            Conversion conversion,
+            TypeSymbol leftType,
+            BoundExpression right,
+            bool isUsed)
+        {
+            if (right.Kind == BoundKind.ConditionalOperator)
+            {
+                var conditional = (BoundConditionalOperator)right;
+                return conditional.Update(
+                    conditional.IsRef,
+                    VisitExpression(conditional.Condition),
+                    RewriteDeconstruction(lhsTargets, conversion, leftType, conditional.Consequence, isUsed: true),
+                    RewriteDeconstruction(lhsTargets, conversion, leftType, conditional.Alternative, isUsed: true),
+                    conditional.ConstantValue,
+                    leftType);
+            }
+
             var temps = ArrayBuilder<LocalSymbol>.GetInstance();
             var effects = DeconstructionSideEffects.GetInstance();
-            ArrayBuilder<Binder.DeconstructionVariable> lhsTargets = GetAssignmentTargetsAndSideEffects(left, temps, effects.init);
-
             BoundExpression returnValue = ApplyDeconstructionConversion(lhsTargets, right, conversion, temps, effects, isUsed, inInit: true);
             effects.Consolidate();
 
-            BoundExpression result;
             if (!isUsed)
             {
                 // When a deconstruction is not used, the last effect is used as return value
                 Debug.Assert(returnValue is null);
                 var last = effects.PopLast();
                 if (last is null)
                 {
-                    // Deconstructions with no effects lower to nothing. For example, `(_, _) = (1, 2);`
-                    result = null;
                     temps.Free();
-                    _ = effects.ToImmutableAndFree();
+                    effects.Free();
```

Comment:
>effects.Free(); [](start = 20, length = 15)

But then if `DeconstructionSideEffects.Free()` properly frees all four arrays, then in this case we'll have a problem (since `Consolidate()` was already called and free'd three of them).
A possible solution is for `Consolidate()` to not free anything (maybe just do a `Clear` on the three arrays it empties) and then `DeconstructionSideEffects.Free()` and `.ToImmutableAndFree()` will be responsible for freeing all four arrays.

Reply:
I don't see the problem. As I said in the above comment, Consolidate is always called before leaving the method. so the only remaining array is `init` that we free here.

Reply:
Ah, good point. We're ok then. Thanks!

---
In reply to: [218484029](https://github.com/dotnet/roslyn/pull/29645#discussion_r218484029) [](ancestors = 218484029)